flink-tests/src/test/java/org/apache/flink/test/checkpointing/ChangelogPeriodicMaterializationITCase.java

```diff
@@ -0,0 +1,206 @@
+                        },
+                        generateJobID()));
+    }
+
+    /** Recovery from checkpoint containing non-materialized state and materialized state. */
+    @Test
+    public void testMaterialization() throws Exception {
+        File checkpointFile = TEMPORARY_FOLDER.newFolder();
+        JobID jobID = generateJobID();
+        SharedReference<AtomicInteger> currentCheckpointNum =
+                sharedObjects.add(new AtomicInteger());
+        SharedReference<List<Long>> currentMaterializationId = sharedObjects.add(new ArrayList<>());
+        StreamExecutionEnvironment env =
+                getEnv(delegatedStateBackend, checkpointFile, 100, 2, 10, 0);
+        waitAndAssert(
+                buildJobGraph(
+                        env,
+                        new ControlledSource() {
+                            @Override
+                            protected void beforeElement(SourceContext<Integer> ctx)
+                                    throws Exception {
+                                if (getRuntimeContext().getAttemptNumber() == 0
+                                        && currentIndex == TOTAL_ELEMENTS / 4) {
+                                    waitWhile(
+                                            () -> {
+                                                if (completedCheckpointNum.get() <= 0) {
+                                                    return true;
+                                                }
+                                                List<Long> allMaterializationId =
+                                                        getAllMaterializationId(
+                                                                checkpointFile, jobID);
+                                                if (!allMaterializationId.isEmpty()) {
+                                                    currentMaterializationId
+                                                            .get()
+                                                            .addAll(allMaterializationId);
+                                                    Collections.sort(
+                                                            currentMaterializationId.get());
+                                                    currentCheckpointNum
+                                                            .get()
+                                                            .compareAndSet(
+                                                                    0,
+                                                                    completedCheckpointNum.get());
+                                                    return false;
+                                                }
+                                                return true;
+                                            });
+
+                                    throwArtificialFailure();
+                                } else if (getRuntimeContext().getAttemptNumber() == 1
```

Comment:
What if some other failure happens and `getAttemptNumber` returns 2?
I guess it's not allowed by maxRestartAttempts == 2, but probably it makes sense to enforce this (e.g. with `Preconditions.checkState`)?